backend/routes.js

```diff
@@ -17,44 +17,67 @@ import {
 import QueryScheduler from './persistent/QueryScheduler.js';
 import {getSetting, saveSetting} from './settings.js';
 import {checkWritePermissions} from './persistent/PlotlyAPI.js';
-import {dissoc, contains, isEmpty, merge, pluck} from 'ramda';
+import {contains, has, isEmpty, merge, pluck} from 'ramda';
+import {getCerts, fetchAndSaveCerts, setRenewalJob} from './certificates';
 import Logger from './logger';
 import fetch from 'node-fetch';
 
+const opn = require('opn');
 const HOSTS = '/etc/hosts';
 
-
-// return HTTPS certs if they exist for a server to use when created or null
-export function getCerts() {
-    try {
-        const keyFile = `${__dirname}/..${getSetting('KEY_FILE')}`;
-        const certFile = `${__dirname}/..${getSetting('CSR_FILE')}`;
-        return {
-            key: fs.readFileSync(keyFile),
-            certificate: fs.readFileSync(certFile)
-        };
-    } catch (e) {
-        return {};
-    }
-}
-
 export default class Server {
     constructor(args = {}) {
         this.certs = getCerts();
         /*
-        // if server is given a protocol argument, enforce it,
-        // otherwise decide which protocol to use
-        // depending if there are certificates
+         * If server is given a protocol argument, enforce it,
+         * otherwise decide which protocol to use
+         * depending if there are certificates. This is mainly
+         * used in tests.
         */
         const apiVersion = '1.0.0';
+        this.apiVersion = apiVersion;
         if (args.protocol === 'HTTP' || isEmpty(this.certs)) {
             this.server = restify.createServer({version: apiVersion});
             this.protocol = 'http';
             this.domain = 'localhost';
+            const redirectBaseUrl = `${this.protocol}://${this.domain}:${getSetting('PORT')}`;
+            const oauthClientId = 'isFcew9naom2f1khSiMeAtzuOvHXHuLwhPsM7oPt';
+            const oauthUrl = (
+                'https://plot.ly/o/authorize/?response_type=token&' +
+                `client_id=${oauthClientId}&` +
+                `redirect_uri=${redirectBaseUrl}/oauth2/callback`
+            );
+            // We wanna skip fetching certs in cases such as tests.
+            if (!args.skipFetchCerts) {
+                const users = getSetting('USERS');
+                /*
+                 * If no USERS in settings we need to go through OAuth before
+                 * being able to ask for a certificate from the CA.
+                 * Let the app load for a while before launching OAuth
+                 * to avoid opening the new tab in the browser without
+                 * the attention of the user.
+                 */
+                if (isEmpty(users)) {
+                    // `opn` module opens a tab in the default browser.
+                    setTimeout(() => opn(oauthUrl), 10000);
```

Comment:
10 seconds feels like a really long time. Is there any other way to trigger this? If we open a new popup window instead of a tab, does that help? Can we trigger this from the front-end instead of the backend like via an explicit button-click?